src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+                $"Attachment {key} has etag {oldEtag}, but Put was called with etag {expectedEtag}. Optimistic concurrency violation, transaction will be aborted.")
+            {
+                ActualETag = oldEtag,
+                ExpectedETag = expectedEtag ?? -1
+            };
+
+                    var streamIdentifier = oldValue.Read(streamIdentifierPosition, out size);
+                    tbv.Add(streamIdentifier, size);
+                    streamIdentifierScope = Slice.External(context.Allocator, streamIdentifier, size, out streamIdentifierSlice);
+
+                    table.Update(oldValue.Id, tbv);*/
+                }
+                else
+                {
+                    if (expectedEtag.HasValue && expectedEtag.Value != 0)
+                    {
+                        ThrowConcurrentExceptionOnMissingAttacment(documentId, name, expectedEtag.Value);
+                    }
+
+                    tbv.Add(newEtag); // streamIdentifier
+                    tbv.SliceFromLocation(context.Allocator, (int) AttachmentsTable.StreamIdentifier, out streamIdentifierSlice);
+
+                    table.Insert(tbv);
+                }
+
+                // Insert the stream
+                var tree = context.Transaction.InnerTransaction.CreateTree(AttachmentsSlice);
+                tree.AddStream(streamIdentifierSlice, stream);
+
+                // Update the document
+                UpdateDocumentAfterAttachmentPut(context, lowerDocumentId, lowerDocumentIdSize, documentId, name, modifiedTicks);
+
+                _documentDatabase.Metrics.AttachmentPutsPerSecond.MarkSingleThreaded(1);
+                _documentDatabase.Metrics.AttachmentBytesPutsPerSecond.MarkSingleThreaded(stream.Length);
+            }
+
+            return newEtag;
+        }
+
+        private void UpdateDocumentAfterAttachmentPut(DocumentsOperationContext context, byte* lowerDocumentId, int lowerDocumentIdSize, 
+            string documentId, string name, long modifiedTicks)
+        {
+            Slice loweredKey;
+            using (Slice.External(context.Allocator, lowerDocumentId, lowerDocumentIdSize, out loweredKey))
+            {
+                var readTable = new Table(DocsSchema, context.Transaction.InnerTransaction);
+                TableValueReader tvr;
+                if (readTable.ReadByKey(loweredKey, out tvr) == false)
+                {
+                    ThrowDocumentNotFound(context, loweredKey, documentId, name);
+                }
+
+                int size;
+                var data = new BlittableJsonReaderObject(tvr.Read((int) DocumentsTable.Data, out size), size, context);
+                var collectionName = ExtractCollectionName(context, documentId, data);
+                var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema, collectionName.GetTableName(CollectionTableType.Documents));
+                if (table.ReadByKey(loweredKey, out tvr) == false)
```

Comment:
I changed it so we update the document before we save the attachment, but still the document etag will be greater than the attachment.